Tasks/ServiceFabricComposeDeploy/ServiceFabricComposeDeploy.ps1

```diff
@@ -163,48 +163,101 @@ try {
+            Remove-ServiceFabricComposeApplicationHelper -ApiVersion $apiVersion -RemoveParameters $removeParameters
+            do
+            {
+                Write-Host (Get-VstsLocString -Key CurrentStatus -ArgumentList $existingApplication.Status)
+                Start-Sleep -Seconds 3
+                $existingApplication = Get-ServiceFabricComposeApplicationStatusHelper -ApiVersion $apiVersion -GetStatusParameters $getStatusParameters
+            }
+            while ($existingApplication -ne $null)
+            Write-Host (Get-VstsLocString -Key ApplicationRemoved)
+        }
+    }
 
-    Write-Host (Get-VstsLocString -Key WaitingForDeploy)
-    $newApplication = Get-ServiceFabricComposeApplicationStatusHelper -ApiVersion $apiVersion -GetStatusParameters $getStatusParameters
-    while (($newApplication -eq $null) -or `
-           ($newApplication.Status -eq 'Provisioning') -or `
-           ($newApplication.Status -eq 'Creating'))
+    if ($upgrading)
     {
-        if ($newApplication -eq $null)
+        $upgradeParameters = New-Object 'System.Collections.Hashtable' $deployParameters
+        $upgradeParameters.Monitored = $True
+        $upgradeParameters.FailureAction = 'Rollback'
+        $upgradeParameters.ConsiderWarningAsError = $True
+        $upgradeParameters.Force = $True
+
+        $upgradeStatus = Get-ServiceFabricComposeDeploymentUpgradeHelper -ApiVersion $apiVersion -GetUpgradeParameters $getStatusParameters
+        if (($upgradeStatus -ne $null) -and (IsUpgradeRunning $upgradeStatus.UpgradeState))
         {
-            Write-Host (Get-VstsLocString -Key WaitingForDeploy)
+            Write-Error (Get-VstsLocString -Key UpgradeInProgress -ArgumentList $applicationName)
         }
-        else
+
+        Write-Host (Get-VstsLocString -Key UpgradingApplication)
+        Start-ServiceFabricComposeDeploymentUpgradeHelper -ApiVersion $apiVersion -UpgradeParameters $upgradeParameters
+
+        Write-Host (Get-VstsLocString -Key WaitingForUpgrade)
+        # Wait a minute before checking on the upgrade to avoid getting the status of the last upgrade.
+        Start-Sleep -Seconds 60
+        $upgradeStatus = Get-ServiceFabricComposeDeploymentUpgradeHelper -ApiVersion $apiVersion -GetUpgradeParameters $getStatusParameters
+        while (($upgradeStatus -eq $null) -or (IsUpgradeRunning $upgradeStatus.UpgradeState))
         {
-            Write-Host (Get-VstsLocString -Key CurrentStatus -ArgumentList $newApplication.Status)
+            if ($upgradeStatus -eq $null)
+            {
+                Write-Host (Get-VstsLocString -Key WaitingForUpgrade)
+            }
+            else
+            {
+                Write-Host (Get-VstsLocString -Key CurrentStatus -ArgumentList $upgradeStatus.UpgradeState )
+            }
+            Start-Sleep -Seconds 3
+            $upgradeStatus = Get-ServiceFabricComposeDeploymentUpgradeHelper -ApiVersion $apiVersion -GetUpgradeParameters $getStatusParameters
```

Comment:
When does $upgradeStatus become null after upgrade is complete? Can it happen in 3 seconds window here?

Reply:
The docs don't give a specific time to live that I've found. But I have accessed the results days later.